### PR TITLE
Add OpenTelemetry support to ON CLUSTER DDL

### PR DESCRIPTION
**Pull request by FrankChen021**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add OpenTelemetry support to ON CLUSTER DDL(require `distributed_ddl_entry_format_version` to be set to 4)

ON CLUSTER DLL is one of the most import statments in ClickHouse. Such statement is scheduled on all nodes in a cluster. It's useful when the OpenTelemetry tracing capability is added to such statements so that we know:
- how many nodes are actually scheduled
- which nodes have been scheduled for one specific DDL
- when the DDL is scheduled on each node
- If it fails, which node fails to execute the DDL

Take following log as an example, we know that node 005 received the DDL, and then the DDL was dispatched to 003/004/005 to execute.

image

This feature has been running well on our clusters for almost 1 year since 21.8.

### CODE REVIEW

**Comment by FrankChen021** [comment]
The new tests fails and the log shows as

```
2022-09-20 13:56:15 02423_ddl_for_opentelemetry:                                            [ FAIL ] 2.41 sec. - result differs with reference: 
2022-09-20 13:56:15 Binary files /ClickHouse/tests/queries/0_stateless/02423_ddl_for_opentelemetry.reference and /ClickHouse/tests/queries/0_stateless/02423_ddl_for_opentelemetry.stdout differ
2022-09-20 13:56:15 
2022-09-20 13:56:15 
2022-09-20 13:56:15 Database: test_a1l1ckyi
```

But the log does not show the diff. Do you know what might be the cause? @tavplubix

**Comment by tavplubix** [comment]
Test have printed something weird to stdout (but I have no ideas why), so `diff` got confused and decided that file is binary. You can add code that checks the output of `diff` command and prints `stdout_file` if `diff` reported that file is binary. Diff is checked here:
https://github.com/ClickHouse/ClickHouse/blob/ecdea62cd5ab23254701d2a89c42c231d51a91da/tests/clickhouse-test#L852-L868

**Comment by FrankChen021** [comment]
The failed tests look like has nothing to do these the changes in the PR. Update the branch th see if it helps.

**Comment by tavplubix** [comment]
>  how can I exclude the new tests from Stateless tests (release, DatabaseReplicated) ? 

Add `no-replicated-database` tag

> Because there're ON CLUSTER DDLs in these cases, and these DDLs are not supported such database engine.

It's because `Replicated` database replicates DDL queries itself (almost the same way as `DDLWorker` does), so `ON CLUSTER` clause is redundant. But if cluster name equals to database name, then `ON CLUSTER` clause is ignored. So I would suggest to use something like
```
cluster_name=$($CLICKHOUSE_CLIENT -q "select if(engine = 'Replicated', name, 'test_shard_localhost')  from system.databases where name='$CLICKHOUSE_DATABASE'")
...
execute_query $trace_id "CREATE TABLE ddl_test_for_opentelemetry ON CLUSTER $cluster_name ...
```
to test that it will work with Replicated database as well

**Comment by FrankChen021** [comment]
> Add `no-replicated-database` tag
> 

I've added such tag to test case, but it's still executed.


> It's because `Replicated` database replicates DDL queries itself (almost the same way as `DDLWorker` does), so `ON CLUSTER` clause is redundant. But if cluster name equals to database name, then `ON CLUSTER` clause is ignored. So I would suggest to use something like
> 
> ```
> cluster_name=$($CLICKHOUSE_CLIENT -q "select if(engine = 'Replicated', name, 'test_shard_localhost')  from system.databases where name='$CLICKHOUSE_DATABASE'")
> ...
> execute_query $trace_id "CREATE TABLE ddl_test_for_opentelemetry ON CLUSTER $cluster_name ...
> ```
> 
> to test that it will work with Replicated database as well

Good suggestion. We have not used the Replicated database engine yet, let me check if the tracing works correctly for such database engine first.

**Comment by FrankChen021** [comment]
> You have added it to another test, see diff

You're right. Such a stupid mistake😄

File: src/Interpreters/DDLTask.cpp [hunk 1]
```diff
@@ -94,6 +94,8 @@ String DDLLogEntry::toString() const
         wb << "settings: " << serializeAST(ast) << "\n";
     }
 
+    wb << this->tracing_context;
+
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply src/Interpreters/DDLTask.cpp:98]: > If you want to use new feature (that may affect compatibility), you explicitly enable a setting. You cannot go back once you enabled new feature and started to use it, but you can if you did not.

Do you mean that you or other people run cluster by explicitely setting  `distributed_ddl_entry_format_version` in their profile? And after upgrading, you change this setting to the newer version? If so, I can understand why the `assertEOF` lays there.

[the remark this answers is not in the document]
Comment (tavplubix) [reply src/Interpreters/DDLTask.cpp:98]: Yes, if they want to use a new feature. But we usually keep compatible default values of such settings for a while, so in most cases users can upgrade/downgrade between LTS releases without explicitly setting something in config. And after 6-12 months a setting for a new feature can be enabled by default (like https://github.com/ClickHouse/ClickHouse/pull/35463)

[the remark this answers is not in the document]
Comment (FrankChen021) [reply src/Interpreters/DDLTask.cpp:98]: Maybe the `client_tcp_protocol_version` I mentioned above is not a good example. 

If we look at how bin files are serialized and deserialized in DistributedSink, it also provides a compatible ways for hosts running in different versions to process the distributed insert correctly during upgrading and downgrading. And we don't need such a user level setting to let operators to configure it.

BTW, considering current use habits, just remove the `asserteEOF` does not help for next upgrade. Whether we need to improve the it and how can we improve it is another topic that is out of the scope of this PR. So, let me follow current patterns to handle the compatibility problem here.

**Review by tavplubix — APPROVED** [review]

**Comment by FrankChen021** [comment]
> I suppose it's because Replicated database creates DDL entry through `DatabaseReplicated::tryEnqueueReplicatedDDL`/`DatabaseReplicatedDDLWorker::tryEnqueueAndExecuteEntry`, not through `executeDDLQueryOnCluster` as usual ON CLUSTER query does. Maybe adding `OpenTelemetry::SpanHolder span(__FUNCTION__);` and `entry.tracing_context = OpenTelemetry::CurrentContext();` to one of these methods will fix it.

Yes, `executeDDLQueryOnCluster` is not executed for Replicated database. And current test cases check this span(check the cluster name kept in the attribute of that span). So I need to change the verification to adapt to Replicated databases.

When I ran the test cases against Replicated database on my local, the cases didn't fail. Maybe there're something wrong with my local configuration.

And `entry.tracing_context = OpenTelemetry::CurrentContext();`  was missed for Replicated database indded. 
But it does not affect the tracing because it calls `DDLWorker::processTask` at last on same stack and when the latter one is called, it tries to set up tracing context, and it finds current thread already has tracing context, so it uses the context on the thread instead of the context in entry. This is how `TracingContextOnThread` designed for such case.

Here is a log for Replicated database when I ran the test case on my local this afternoon.

image

**Comment by FrankChen021** [comment]
Interesting, the ON CLUSTER DDL output it's execution status to stdout while I explicitly supress it by given `distributed_ddl_output_mode=none`. And such execution status is not printed on my local.

**Comment by FrankChen021** [comment]
The `Stateless tests (release, DatabaseReplicated)` still fails. The first DDL threw exception while the other DDLs were executed successfully.

```
2022.09.24 03:19:16.208831 [ 664 ] {} <Trace> DynamicQueryHandler: Request URI: /?max_insert_threads=6&group_by_two_level_threshold=100000&group_by_two_level_threshold_bytes=50000000&distributed_aggregation_memory_efficient=0&fsync_metadata=0&output_format_parallel_formatting=0&input_format_parallel_parsing=1&min_chunk_bytes_for_parallel_parsing=21409094&max_read_buffer_size=582693&prefer_localhost_replica=1&max_block_size=8857&max_threads=48&optimize_or_like_chain=1&optimize_read_in_order=0&read_in_order_two_level_merge_threshold=41&optimize_aggregation_in_order=0&aggregation_in_order_max_block_bytes=31524761&use_uncompressed_cache=0&min_bytes_to_use_direct_io=0&min_bytes_to_use_mmap_io=0&local_filesystem_read_method=pread_threadpool&remote_filesystem_read_method=threadpool&local_filesystem_read_prefetch=1&remote_filesystem_read_prefetch=0&compile_expressions=0&compile_aggregate_expressions=1&compile_sort_description=0&merge_tree_coarse_index_granularity=5&optimize_distinct_in_order=0&optimize_sorting_by_input_stream_properties=1&database=test_b6b84g3f&log_comment=02423_ddl_for_opentelemetry.sh&distributed_ddl_output_mode=none&distributed_ddl_entry_format_version=1
2022.09.24 03:19:16.209273 [ 664 ] {0540d831-eab2-42bc-9c38-08f5efffdd90} <Debug> executeQuery: (from [::1]:38966) (comment: 02423_ddl_for_opentelemetry.sh) CREATE TABLE test_b6b84g3f.ddl_test_for_opentelemetry ON CLUSTER test_b6b84g3f (id UInt64) Engine=MergeTree ORDER BY id (stage: Complete)

...


2022.09.24 03:20:18.695933 [ 664 ] {0540d831-eab2-42bc-9c38-08f5efffdd90} <Error> executeQuery: Code: 341. DB::Exception: Cannot provide query execution status. The query's node /test/clickhouse/db/test_b6b84g3f/log/query-0000000004 has been deleted by the cleaner since it was finished (or its lifetime is expired). (UNFINISHED) (version 22.10.1.123) (from [::1]:38966) (comment: 02423_ddl_for_opentelemetry.sh) (in query: CREATE TABLE test_b6b84g3f.ddl_test_for_opentelemetry ON CLUSTER test_b6b84g3f (id UInt64) Engine=MergeTree ORDER BY id), Stack trace (when copying this message, always include the lines below):
```

Not sure why such exception is raised. @tavplubix Could you help take a look?

Another question is that, the above DDL was returned at 
2022.09.24 03:20:**18**.695933
however, the queries after the DDL in the test case file were started to execute at  
2022.09.24 03:20:**16**.253178
How could it be? Since prev queries were not finished yet, span logs were not written, the verification in the case failed.

```
2022.09.24 03:19:16.209273 [ 664 ] {0540d831-eab2-42bc-9c38-08f5efffdd90} <Debug> executeQuery: (from [::1]:38966) (comment: 02423_ddl_for_opentelemetry.sh) CREATE TABLE test_b6b84g3f.ddl_test_for_opentelemetry ON CLUSTER test_b6b84g3f (id UInt64) Engine=MergeTree ORDER BY id (stage: Complete)
2022.09.24 03:20:16.253178 [ 1628 ] {de14585f-efe2-4217-959f-d21babae6322} <Debug> executeQuery: (from [::1]:44958) (comment: 02423_ddl_for_opentelemetry.sh) SYSTEM FLUSH LOGS; (stage: Complete)
```

**Comment by tavplubix** [comment]
> Not sure why such exception is raised.

Because the database was being removed concurrently:
```
2022.09.24 03:20:18.219910 [ 666 ] {89670df1-78b0-44dc-a70a-44df7ca0de09} <Debug> executeQuery: (from [::1]:40334) (comment: 02423_ddl_for_opentelemetry.sh) DROP DATABASE test_b6b84g3f ON CLUSTER test_cluster_database_replicated  (stage: Complete)
2022.09.24 03:20:18.219937 [ 666 ] {89670df1-78b0-44dc-a70a-44df7ca0de09} <Trace> ContextAccess (default): Access granted: CLUSTER ON *.*
2022.09.24 03:20:18.219948 [ 666 ] {89670df1-78b0-44dc-a70a-44df7ca0de09} <Trace> ContextAccess (default): Access granted: DROP DATABASE ON test_b6b84g3f.*
2022.09.24 03:20:18.323192 [ 666 ] {89670df1-78b0-44dc-a70a-44df7ca0de09} <Information> executeQuery: Read 3 rows, 159.00 B in 0.10325691 sec., 29 rows/sec., 1.50 KiB/sec.
2022.09.24 03:20:18.323282 [ 666 ] {89670df1-78b0-44dc-a70a-44df7ca0de09} <Debug> DynamicQueryHandler: Done processing query
2022.09.24 03:20:18.323297 [ 666 ] {89670df1-78b0-44dc-a70a-44df7ca0de09} <Debug> MemoryTracker: Peak memory usage (for query): 1.03 MiB.
```

So all its nodes (including DDL queue and `/test/clickhouse/db/test_b6b84g3f/log/query-0000000004`) were removed. `DROP DATABASE` is executed by `clickhouse-test` after test is finished.

Seems like default timeout for `curl` is not enough, so `curl` exits before query is actually finished. And it does not show any error messages because of `-f` flag. It also answers the second question: 
> How could it be? Since prev queries were not finished yet, span logs were not written, the verification in the case failed.

It's because `curl` does not wait for prev query to finish

**Comment by FrankChen021** [comment]
Thank you @tavplubix very much for improving the test cases

**Comment by tavplubix** [comment]
AST fuzzer (asan) - https://github.com/ClickHouse/ClickHouse/issues/41822
Stress test (asan) - https://github.com/ClickHouse/ClickHouse/issues/41104
Stress test (debug) - https://github.com/ClickHouse/ClickHouse/pull/41717
Stress test (msan) - https://github.com/ClickHouse/ClickHouse/issues/41831
Stress test (tsan) - https://github.com/ClickHouse/ClickHouse/issues/41783
Stress test (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/41729

File: tests/queries/0_stateless/02423_ddl_for_opentelemetry.sh [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+# $1 - OpenTelemetry Trace Id
+# $2 - Query
+# $3 - Query Settings
+function execute_query()
+{
+    # Some queries are supposed to fail, use -f to suppress error messages
+    echo $2 | ${CLICKHOUSE_CURL_COMMAND} -q -s --max-time 180 \
+                -X POST \
+                -H "traceparent: 00-$1-5150000000000515-01" \
+                -H "tracestate: a\nb cd" \
```
Comment (azat) [inline tests/queries/0_stateless/02423_ddl_for_opentelemetry.sh:22]: FWIW this is not a test for new line, for this you need to use `$'foo\nbar` syntax, but you cannot pass new line in headers, I've converted the test to `clickhouse-client` in fix for escaping - #50045

**Pull request merged**